plugins/pulp_rpm/plugins/importers/yum/sync.py

```diff
@@ -127,70 +168,91 @@ def run(self):
+                    if not (skip or self.skip_repomd_steps):
+                        self.get_errata(metadata_files)
+
+                with self.update_state(self.progress_status['comps']) as skip:
+                    if not (skip or self.skip_repomd_steps):
+                        self.get_comps_file_units(metadata_files, group.process_group_element,
+                                                  group.GROUP_TAG)
+                        self.get_comps_file_units(metadata_files, group.process_category_element,
+                                                  group.CATEGORY_TAG)
+                        self.get_comps_file_units(metadata_files, group.process_environment_element,
+                                                  group.ENVIRONMENT_TAG)
+
+            except CancelException:
+                report = self.sync_conduit.build_cancel_report(self._progress_summary,
+                                                               self.progress_status)
+                report.canceled_flag = True
+                return report
+
+            except Exception, e:
+                # Check if the caught exception indicates that the mirror is bad.
+                # Try next mirror in the list without raising the exception.
+                # In case it was the last mirror in the list, raise the exception.
+                # In case other exceptions were caught that are not related to the state of the
+                # mirror, raise the exception immediately and do not iterate throught the rest
+                # of mirrors.
+                bad_mirror_exceptions = [error_codes.RPM1004, error_codes.RPM1006]
+                if isinstance(e, PulpCodedException) and \
+                    (e.error_code in bad_mirror_exceptions) and \
+                        url_count != len(self.sync_feed):
+                            continue
+                else:
+                    for step, value in self.progress_status.iteritems():
+                        if value.get('state') == constants.STATE_RUNNING:
+                            value['state'] = constants.STATE_FAILED
+                            value['error'] = str(e)
+                    self.set_progress()
+                    if isinstance(e, PulpCodedException):
+                        raise
+                    report = self.sync_conduit.build_failure_report(self._progress_summary,
+                                                                    self.progress_status)
+                    return report
+
+            finally:
+                # clean up whatever we may have left behind
+                shutil.rmtree(self.tmp_dir, ignore_errors=True)
+
+            self.save_repomd_revision()
+            _logger.info(_('Sync complete.'))
+            return self.sync_conduit.build_success_report(self._progress_summary,
+                                                          self.progress_status)
```

Comment:
This method was already far too complex, and this change is making it even more complex and difficult to read. I think it would be really good to break the above out into small logically separated steps, so that the logic is much easier to follow. It's difficult to tell which code is at which indentation level as it is!